### ISSUE TITLE
request for optimizing journal file size

### ISSUE THREAD

**Issue by ChALkeR** [systemd/systemd#5285]
### Submission type

  - [ ] Bug report
  - [x] Request for enhancement (RFE)

### systemd version the issue has been seen with

> 232, 215

### Used distribution

> Arch Linux (232), Debian Jessie (215)

### To reproduce:
```console
# du -bs /var/log/journal/path-to.journal
134217728       /var/log/journal/path-to.journal
# journalctl --file=/var/log/journal/path-to.journal -o json | wc
126785 12337277 154364438
# journalctl --file=/var/log/journal/path-to.journal -o json | sed -E s/'"[^"]+" : '//g | wc
126785 4985922 92505495
# journalctl --file=/var/log/journal/path-to.journal -o json | sed -E s/'"(__CURSOR|_BOOT_ID|_MACHINE_ID)" : "[^"]+", '//g | sed -E s/'"[^"]+" : '//g | wc
 126785 4605567 67076269
```

It can be seen that journald files on disk are about the same size as fully serialized json output, with field identifiers on each line and all timestamps, identifiers, and cursors serialized to strings.

That's about 1058 bytes per entry for that file (on a desktop). On a server, that's 765 bytes per entry.

JSON output with stripped field identifiers (but still with cursors/ids/timestamps serialized as strings) is about 30% smaller than journald storage size on-disk.

The same output in text format with stripped field identifiers and stripped `__CURSOR`, `_BOOT_ID` and `_MACHINE_ID` fields is _more than two times smaller_ than binary format which journald uses to store logs on disk.

That's not what one would expect from binary logs.

hexdump of `*.journal` files also looks strange:
```
0037d0b0  28 ff fb 03 00 00 00 00  28 ff fb 03 00 00 00 00  |(.......(.......|
0037d0c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0037d0e0  40 68 26 04 00 00 00 00  40 68 26 04 00 00 00 00  |@h&.....@h&.....|
0037d0f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0037d120  e8 f1 d5 03 00 00 00 00  c0 df 27 06 00 00 00 00  |..........'.....|
0037d130  10 97 87 01 00 00 00 00  d0 1b 56 07 00 00 00 00  |..........V.....|
0037d140  f0 9a 29 01 00 00 00 00  b0 3e f4 06 00 00 00 00  |..)......>......|
0037d150  98 b0 37 04 00 00 00 00  98 b0 37 04 00 00 00 00  |..7.......7.....|
0037d160  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0037d170  f8 b3 55 04 00 00 00 00  f8 b3 55 04 00 00 00 00  |..U.......U.....|
0037d180  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0037d190  40 3c 12 07 00 00 00 00  40 3c 12 07 00 00 00 00  |@<......@<......|
0037d1a0  00 d1 6c 02 00 00 00 00  28 10 3d 04 00 00 00 00  |..l.....(.=.....|
0037d1b0  c0 ee 5c 01 00 00 00 00  b8 21 85 01 00 00 00 00  |..\......!......|
0037d1c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0037d1e0  40 e7 65 03 00 00 00 00  40 e7 65 03 00 00 00 00  |@.e.....@.e.....|
0037d1f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0037d200  00 7b 68 06 00 00 00 00  00 7b 68 06 00 00 00 00  |.{h......{h.....|
0037d210  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0037d230  68 53 a5 02 00 00 00 00  68 53 a5 02 00 00 00 00  |hS......hS......|
0037d240  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0037d280  a0 cc 64 03 00 00 00 00  a0 cc 64 03 00 00 00 00  |..d.......d.....|
0037d290  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0037d2d0  a0 6a b1 07 00 00 00 00  a0 6a b1 07 00 00 00 00  |.j.......j......|
0037d2e0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0037d300  98 b3 4c 04 00 00 00 00  98 b3 4c 04 00 00 00 00  |..L.......L.....|
0037d310  58 26 82 02 00 00 00 00  58 26 82 02 00 00 00 00  |X&......X&......|
```

That's not a special case — log files look like that most of the way through them.

I am not sure if this a bug in the implementation or an actual design decision.
If the latter, there could be several ways to improve that, from using 32-bit offsets instead of 64-bit ones if log is split into files smaller than 4 GiB (`SystemMaxFileSize` < 4 GiB, that seems to be true by default) to utilizing chunked file compression with lz4 (like fs compression works, that would still allow random access).

**Comment by poettering** [comment]
Uh? the fact that the format is binary has nothing to do with making it small, it has all to do with indexing it. We do compress overly large objects stored in it, but only those.

And I am sorry, but 32bit offsets isn't really an option today. THis is not a smart way to save disk space, really.

Anyway, there's certainly room for improvement, but the current behaviour is very much within what is to be expected.

**Comment by intelfx** [comment]
@poettering I suppose the main question is "why the journal files look so sparse". Given that there is field-level deduplication built in the file format, it is pretty unexpected that a journal file still takes more space than its full serialization.

**Comment by ChALkeR** [comment]
> the fact that the format is binary has nothing to do with making it small, it has all to do with indexing it. 

True that binary ≠ small, but having a binary format means that some methods of saving the disk space could be utilized (like deduplication and compression), and because of that users generally don't expect binary logs to be more than two times larger than text logs.

> And I am sorry, but 32bit offsets isn't really an option today. THis is not a smart way to save disk space, really.

I might be misunderstanding something, but what specific usecases would adding support for 32-bit offsets for files that are smaller than 4 GiB affect in a negative way? The smart way to save disk space would be to add chunked compression for the whole file, but that would probably require significantly larger changes.

One could argue that this is the responsibility of the fs and that users should just use file systems with built-in compression for log files, but for most setups out there fs is not compressed.

**Comment by Glandos** [comment]
```
-rw-r----- 1 root root 109051904 Feb 26 14:39 system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal
-rw-r--r-- 1 root root  24565838 Feb 26 16:50 system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal.lz4
-rw-r----- 1 root root  13119374 Feb 26 14:39 system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal.zst
```

Here are the commands that generate this:
```
> time lz4 -f system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal
Compressed filename will be : system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal.lz4 
Compressed 109051904 bytes into 24565838 bytes ==> 22.53%                      
1.02user 0.31system 0:01.34elapsed 99%CPU (0avgtext+0avgdata 7548maxresident)k
0inputs+47984outputs (0major+1588minor)pagefaults 0swaps

> time zstd -f -1 system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal 
system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal : 12.03%   (109051904 => 13119374 bytes, system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal.zst) 
1.84user 0.22system 0:02.07elapsed 99%CPU (0avgtext+0avgdata 3208maxresident)k
0inputs+25624outputs (0major+406minor)pagefaults 0swaps
```

My point is that, on top of internal compression (which is enabled on that example) a simple lz4 compression give a file that is 22,5% of the initial one, and if you are willing to use zstd, it can be better, with a slight increase of CPU time.
And decompression is quite fast:
```
> time lz4 -f -d system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal.lz4
Decoding file system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal 
Successfully decoded 109051904 bytes                                           
0.43user 0.83system 0:01.61elapsed 78%CPU (0avgtext+0avgdata 7644maxresident)k
48inputs+178608outputs (0major+1632minor)pagefaults 0swaps


> time zstd -f -d system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal.zst
system@7615a7879fce4188bd6d29c5f222673b-0000000002e93b8b-00055d03c1e042f6.journal.zst: 109051904 bytes 
0.76user 0.76system 0:01.53elapsed 99%CPU (0avgtext+0avgdata 2716maxresident)k
0inputs+178608outputs (0major+321minor)pagefaults 0swaps
```

Tests were done on a Intel(R) Atom(TM) CPU D2550   @ 1.86GHz. Not really powerful.

**Comment by ZNikke** [comment]
Having the archived log files compressed would actually be what I naively expect when journald compression is enabled. It was quite surprising realizing that the log file was very sparse with the current way of doing things.

I for one would be in favour for just compressing archived log files with lz4/zstd/whatever, keeping the .lz4/.zstd/.etc file ending and just adding support in journalctl to handle them. I could live with systemctl et al not being able to view them, especially if there was a smooth way to specify how many archived log files to keep before whole-file-compressing them.

One (of many) caveats would be that you'd likely want smaller log file size defaults to avoid having too large uncompressed log file blobs in memory when processing files.

As it stands, the systemd journals are less useful on disk-constrained systems (read: virtual machines where large unique logs multiplied with many hosts means substantial amounts of data). Text logs that are compressed enables keeping much longer history within the same storage envelope.

**Comment by cedws** [comment]
This is quite a problem. I recently moved a few services over to logging purely to systemd, naively assuming journald would optimise and compress logs on disk for me. The journal is now filling up way too fast and storage is at a premium in this particular environment, so increasing journal size is not an option.
Why does it appear that journald is only using block compression? Wouldn't streaming compression be more suitable?
Could really do with an option for writing the journal to a custom path too, currently doing this with a bind mount...

**Comment by tobizviz** [comment]
I encountered this issue because I am using memory only logging, and wend down the rabbit hole trying to figure out how it could be possible that journald uses hundreds of times more space than legacy rsyslog text files to store essentially the same amount of data.

My system is memory constrained with a read-only filesystem, so I am forced to switch to rsyslog for storage of logs. I'll set up forwarding from journald to rsyslog and allocate a minimal amount of ephemeral storage for journald.

**Comment by wavexx** [comment]
@tobizviz For small systems I forward to rsyslog for exactly this reason. For archival I keep ~1w work of logs on journald, then I also forward to rsyslog where for the same space I can keep 1 year of data. I wished this wasn't necessary.

**Comment by xloem** [comment]
Just noting that writing a transparent file interface for a compressed format is not very much work at all.  It should be pretty reasonable for the community to offer a fork with a solution here.

**Comment by socketpair** [comment]
Please see how compression and indexing work at the same time in ClickHouse. Its compression is perfect, and structure is data semantics is the same as in Journald. ClickHouse is a column-oriented database with append-only semantics. Especially, it does its best on low cardinality columns.

Another option - is to use  RocksDB as low level storage format. It is very fast and compact and the same time.

**Comment by socketpair** [comment]
@poettering what do you think about investigation of ClickHouse storage ideas and also about RocksDB ?

**Comment by mrc0mmand** [comment]
Please see recent @DaanDeMeyer's work regarding the journal's compact mode - https://github.com/systemd/systemd/pull/21183 - which should address some of the concerns here.

**Comment by DaanDeMeyer** [comment]
To expand on @mrc0mmand's reply, I've been working on a few improvements for journal size:

- BTRFS filesystem compression will be used for archived journal files in v250 if enabled using the compress mount option. Archived journal files compress to roughly 13% of their original size.
- Archived journal files will be truncated and holes will be punched in unused parts of the file in the next release. This results in some good savings as well (10% reduction)
- https://github.com/systemd/systemd/pull/21183 will roughly halve journal file size when it lands

I'm also planning to add support for compressing archived journal files (for cases where filesystem compression is not available) which should provide further wins.

**Comment by Glandos** [comment]
> To expand on @mrc0mmand's reply, I've been working on a few improvements for journal size:
> 
>     * BTRFS filesystem compression will be used for archived journal files in v250 if enabled using the compress mount option. Archived journal files compress to roughly 13% of their original size.

BTRFS is a good filesystem, but really heavy on resources. I a have a lot of low-end computers and it's not really practical to use BTRFS on them.

>     * Archived journal files will be truncated and holes will be punched in unused parts of the file in the next release. This results in some good savings as well (10% reduction)
> 
>     * [journal: Add compact mode  #21183](https://github.com/systemd/systemd/pull/21183) will roughly halve journal file size when it lands
> 
> 
> I'm also planning to add support for compressing archived journal files (for cases where filesystem compression is not available) which should provide further wins.

That's really good news!

**Comment by linuxtim** [comment]
>     * [journal: Add compact mode  #21183](https://github.com/systemd/systemd/pull/21183) will roughly halve journal file size when it lands

To update this, a subset of #21183 was merged as PR #22998, which was first included in systemd release version 252.  The compact storage format is enabled by default in version 252 or later (but can be disabled by running journald or journalctl with the environmental variable `SYSTEMD_JOURNAL_COMPACT` set to `off` or equivalent).

**Comment by dechamps** [comment]
Yeah ZFS compression works wonders for the journal and that's what I used as well, but if you use that approach then be aware of #18353: unfortunately jourrnald gets confused about how ZFS reports file sizes when using compression so it needs a bit of additional tuning if you don't want to add up with lots and lots of journal files (which slows down journal reads considerably, IIRC).

**Comment by poettering** [comment]
Let's close this, as https://github.com/systemd/systemd/pull/22998 has long been merged.

**Comment by tim-seoss** [comment]
@DaanDeMeyer Thanks for your work on this issue.  Are there any relevant issues or PRs which covers the remaining non-merged features from #21183 - which would help people who were tracking this bug follow continued progress on journal file size optimisation?  As it stands journald persistent mode is not usable in many environments due to the significant increase in log file size vs. older options like rsyslog + logrotate...
